db-tests/src/test/java/gov/nasa/jpl/aerie/database/ExternalEventTests.java

```diff
@@ -0,0 +1,2210 @@
+        ExternalSource B = new ExternalSource(
+            "B",
+            SOURCE_TYPE,
+            DERIVATION_GROUP,
+            "2024-01-02T00:00:00Z",
+            "2024-01-01T03:00:00Z",
+            "2024-01-01T04:00:00Z",
+            CREATED_AT
+        );
+        ExternalSource C = new ExternalSource(
+            "C",
+            SOURCE_TYPE,
+            DERIVATION_GROUP,
+            "2024-01-03T00:00:00Z",
+            "2024-01-01T00:00:00Z",
+            "2024-01-01T01:00:00Z",
+            CREATED_AT
+        );
+
+        // negated by empty space
+        ExternalEvent e1 = new ExternalEvent(
+            "a",
+            "2024-01-01T01:50:00Z",
+            "00:10:00",
+            A
+        );
+        // negated by empty space
+        ExternalEvent e2 = new ExternalEvent(
+            "a",
+            "2024-01-01T03:40:00Z",
+            "00:15:00",
+            B
+        );
+        // negated by empty space
+        ExternalEvent e3 = new ExternalEvent(
+            "a",
+            "2024-01-01T00:30:00Z",
+            "00:20:00",
+            C
+        );
+
+        // insert sources
+        insertExternalSource(A);
+        insertExternalSource(B);
+        insertExternalSource(C);
+
+        // insert events
+        insertExternalEvent(e1);
+        insertExternalEvent(e2);
+        insertExternalEvent(e3);
+
+        // verify expected keys
+        final var results = getDerivedEvents();
+
+        // this range should only have a single element
+        final List<DerivedEvent> expectedResults = List.of(
+            new DerivedEvent(
+                "a",
+                SOURCE_TYPE,
+                "C",
+                DERIVATION_GROUP,
+                "2024-01-01 00:30:00+00",
+                "00:20:00",
+                "{[\"2024-01-01 00:00:00+00\",\"2024-01-01 01:00:00+00\")}",
+                "2024-01-03 00:00:00+00"
+            )
+        );
+        assertEquals(expectedResults.size(), results.size());
+        assertTrue(results.containsAll(expectedResults));
+      }
+    }
+  }
+  //endregion
+
+
+  //region Constraint Tests
+  /**
+   * The focus of this class is to provide a final set of tests to test primary and significant constraints to prove
+   *    that they work and how they work. These constraints are baked into the schema, so extensive tests of foreign or
+   *    unique keys are not excluded except for common use cases/easily made mistakes, as PSQL guarantees functionality
+   *    for us.
+   * We seek to demonstrate in this class, and verify as a sanity check.
+   */
+  @Nested
+  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+  class ConstraintTests {
+
+    /**
+     * Generally, no two sources in a derivation group, regardless of a lack of overlap, can have the same valid_at. We
+     *    demonstrate this with sparse sources. In this test, we don't even require events to demonstrate this behavior.
+     */
+    @Test
+    void sameValid_at() throws SQLException {
```

Comment:
I don't think a test of the `unique` constraint placed on the `external_sources` table is meaningful, as it is testing Postgres's internal code and not Aerie logic.